Argparse v3 - RC1

oam7575 opened the pull request:
Version 3 of the argparse change.

Issue : #297 
Implement argparse
Remove getopts

ikwyl6 reviewed (commented):
Hi @oam7575 - this is a lot better. This provides all the changes to one file that is different when compared to the ‘main’ version on jtesta. 
I’m on iPhone now but I think you may want to add some ‘help’ directives to each argparse line item. This is for when you do a ‘ssh_audit -h’ and the ‘help’ option for each argparse line item that will explain all the info for each option you are specifying..

oam7575 commented:
Hey @ikwyl6

I did think of that but deliberately _did not_ put any help dialog in there as ssh-audit already has an info / help block that gets printed out.

As you have seen I was having a few issues and didn't want to make it potentially worse.

With that said; once Joe has time to review this iteration, I would be happy to update the argparse based on his and your feedback.

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -836,82 +839,129 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
+    parser.add_argument('--threads', action="store", dest='threads', default='1', type=int)
+
+    try:
+        namespace = parser.parse_args()
```
Calling this variable `args` might improve readability.

oam7575 replied:
Updating code

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -836,82 +839,129 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
+        aconf.client_audit = namespace.client_audit
+
+        ttime = namespace.timeout
+        if ttime != 5:
```
Hmm... the original code didn't include this logic.  What was the rationale for this?

oam7575 replied:
**Note:** As has been seen I am not well versed with git / merging / comments .
As such the following comments have all been done manually ; so at least I know what I am doing.

The original getopts code has the following 

**Line ~879**
```
        elif o in ('-t', '--timeout'):
            aconf.timeout = float(a)
            aconf.timeout_set = True
```

**Line ~1200**

`s = SSH_Socket(out, aconf.host, aconf.port, aconf.ip_version_preference, aconf.timeout, aconf.timeout_set)`

This passes through and --timeout argument for the test to run.

During my cut over and testing for argparse I encountered several issues were the timeout was populated by a default value.

From memory the error code mentioned a 5 second default test timeout.

My argparse changes forces the the argument to a default value of 5 to match the behavior I observed; checks to see if the value has been modified and if so; sets the timeout value to that number as a float and sets timeout_set = True the same as the original script.

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -836,82 +839,129 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
+
+        ttime = namespace.timeout
+        if ttime != 5:
+            aconf.timeout = float(namespace.timeout)
```
I suppose the `timeout` arg can be defined as a float in the `parser.add_argument()` function, above.

oam7575 replied:
I didn't think to try setting the add_argument as a float vs int.

I was using int's for other numbers and the original code converted the --timeout XX number to a float as seen above.

As you say this could be modified if desired.

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -836,82 +839,129 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
-        elif o in ('-M', '--make-policy'):
+
+        # Get error level regex
+        err_level = (namespace.level)
```
Parentheses wouldn't be needed here.

oam7575 replied:
Ah, that was left over from testing different ways to access the arguments 

similar to :  getattr(arguments, 'list_policies')

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -836,82 +839,129 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
+
+        # Get error level regex
+        err_level = (namespace.level)
+        pattern = re.compile(r'info|warn|fail')
```
An easier method for checking this would be: `if err_level in ["info", "warn", "fail"]:`

oam7575 replied:
Agree - and fairly sure I did attempt that; but was having difficulty with the string matching not working correctly for unknown reasons.

Reasons being either me ; or argparse weirdness that I don't understand.

I'll try again and see if I can make this cleaner

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -836,82 +839,129 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
+            aconf.target_file = namespace.targets
+
+        if os.name == 'nt':
+            aconf.threads = '1'
```
I don't think restricting the number of threads is necessary for Windows platforms.

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -836,82 +839,129 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
+            aconf.threads = '1'
+        
+        if os.name == 'posix':
+            aconf.threads = int(namespace.threads)
```
The `threads` arg is already defined as being an `int`, so no need to call `int()`.

oam7575 replied:
changed

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -836,82 +839,129 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
+        if os.name == 'posix':
+            aconf.threads = int(namespace.threads)
+        else:
+            aconf.threads = '1'
```
I don't think restricting the number of threads here is necessary, either.

oam7575 replied:
From memory this had to do with a threading bug on windows.

Looking at the original code the `inda` (?? ) single thread work around was disabled; and I have implemented a change to force single thread on Windows.

I'll swap this out to parse any --thread value into any OS.

If it works on Windows - great ! ; if it breaks - at least we will have somewhere to look.

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -947,19 +1002,21 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
+    if aconf.client_audit is None and aconf.target_file is None:
         if oport is not None:
-            host = args[0]
+            host = host
```
`host = host` is not needed.

oam7575 replied:
Of all of your comments these are the three I was expecting the most.

Each of the host = host ; port = port  was  me leaving the original code as untouched as possible; where ever it did not need to be touched.

Agree that this section can probably be adjusted and cleaned up.

The original host = args[0] was to take the trailing IP / hostname from the command line.
As I added the --ip arg to ensure I could trap the target IP I left the host =  and just changed it to host = host ; where the host variable is populated on line ~887

These could probably be changed to directly use the arguments similar to 

host = arguments.host
port = arguments.port

again this would be in an attempt to ensure minimal changes and unnecessary breakage on my part

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -947,19 +1002,21 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
-            host, port = Utils.parse_host_and_port(args[0])
-        if not host and aconf.target_file is None:
+            host = host
+            port = port
```
`host = host` and `port = port` isn't needed.

oam7575 replied:
see above

jtesta commented on src/ssh_audit/ssh_audit.py:
```diff
@@ -947,19 +1002,21 @@ def process_commandline(out: OutputBuffer, args: List[str], usage_cb: Callable[.
             port = 2222
         else:
-            port = 22
+            port = port
```
`port = port` isn't needed.

oam7575 replied:
see above

jtesta reviewed (changes requested):
Thanks for this updated PR!  I've left feedback on some code changes.  Once you've made them, I'll start testing it.

jtesta commented:
Additionally, you may find other issues to fix if you run the Tox tests.  A fully passing set of tests isn't necessary for me to merge a PR, though it would make things easier for me.  You can run them with `python3 -m tox` in the top-level directory.

oam7575 commented:
Hi Joe,

I have attempted to fix all the issues you raised; and have run the Tox test.

Please be aware the test_auditconf pytest dies due to Argparse failing on a pytest argument
`SSH Audit Tool: error: unrecognized arguments: --junitxml=/mnt/h/GIT/ssh-audit/reports/junit.py39-test.xml test\n'`

Have also removed references to the `oport` variable, seems that it was used when the SSH port argument was optional.
I have set the default port number to be 22 and updated the code where needed.

Believe it should be working. 

**Note**
I have attempted to make a squash commit to update my original pull request ; but as has been seen previously - I am still getting my head around how to do that cleanly.

If this is a mess - please let me know and I will close this off and create a clean branch / PR with the current code for you to review.

jtesta commented:
@oam7575 : Thanks for working on this!  I applied some additional fixes in a01baadfa8b023abffb18a42e79f78621388fd79 to restore some of the original functionality that was missing, so I gave you partial credit for your work in the change log.  Speaking of which, I could only include your GitHub username since your profile doesn't mention your real name; if you'd like me to list your name instead, please let me know.

Thanks again!

oam7575 commented:
Hi Joe,

Happy with the partial credit / additional fixes. As has been seen, I am still very much learning as I go.

Also happy with the attribution via GitHub Name / Handle.

Cheers.
